Description: fix build against nettle >= 4.0
 nettle 4.0 dropped the digest length argument from the *_digest
 functions. Pick the matching macro form via NETTLE_VERSION_MAJOR so
 both old and new nettle keep building.
Author: Kel Modderman <kel@kodi.tv>
Forwarded: no
Last-Update: 2026-09-02
--- a/librtmp/handshake.h
+++ b/librtmp/handshake.h
@@ -45,6 +45,7 @@
 #elif defined(USE_GNUTLS)
 #include <nettle/hmac.h>
 #include <nettle/arcfour.h>
+#include <nettle/version.h>
 #ifndef SHA256_DIGEST_LENGTH
 #define SHA256_DIGEST_LENGTH	32
 #endif
@@ -52,7 +53,11 @@
 #define HMAC_CTX	struct hmac_sha256_ctx
 #define HMAC_setup(ctx, key, len)	hmac_sha256_set_key(&ctx, len, key)
 #define HMAC_crunch(ctx, buf, len)	hmac_sha256_update(&ctx, len, buf)
+#if NETTLE_VERSION_MAJOR >= 4
+#define HMAC_finish(ctx, dig, dlen)	dlen = SHA256_DIGEST_LENGTH; hmac_sha256_digest(&ctx, dig)
+#else
 #define HMAC_finish(ctx, dig, dlen)	dlen = SHA256_DIGEST_LENGTH; hmac_sha256_digest(&ctx, SHA256_DIGEST_LENGTH, dig)
+#endif
 #define HMAC_close(ctx)
 
 typedef struct arcfour_ctx*	RC4_handle;
--- a/librtmp/hashswf.c
+++ b/librtmp/hashswf.c
@@ -44,6 +44,7 @@
 #define HMAC_close(ctx)
 #elif defined(USE_GNUTLS)
 #include <nettle/hmac.h>
+#include <nettle/version.h>
 #ifndef SHA256_DIGEST_LENGTH
 #define SHA256_DIGEST_LENGTH	32
 #endif
@@ -51,7 +52,11 @@
 #define HMAC_CTX	struct hmac_sha256_ctx
 #define HMAC_setup(ctx, key, len)	hmac_sha256_set_key(&ctx, len, key)
 #define HMAC_crunch(ctx, buf, len)	hmac_sha256_update(&ctx, len, buf)
+#if NETTLE_VERSION_MAJOR >= 4
+#define HMAC_finish(ctx, dig, dlen)	dlen = SHA256_DIGEST_LENGTH; hmac_sha256_digest(&ctx, dig)
+#else
 #define HMAC_finish(ctx, dig, dlen)	dlen = SHA256_DIGEST_LENGTH; hmac_sha256_digest(&ctx, SHA256_DIGEST_LENGTH, dig)
+#endif
 #define HMAC_close(ctx)
 #else	/* USE_OPENSSL */
 #include <openssl/ssl.h>
--- a/librtmp/rtmp.c
+++ b/librtmp/rtmp.c
@@ -57,6 +57,7 @@
 #define MD5_DIGEST_LENGTH 16
 #include <nettle/base64.h>
 #include <nettle/md5.h>
+#include <nettle/version.h>
 #else	/* USE_OPENSSL */
 #include <openssl/ssl.h>
 #include <openssl/rc4.h>
@@ -2492,7 +2493,11 @@
 typedef struct md5_ctx	MD5_CTX;
 #define MD5_Init(ctx)	md5_init(ctx)
 #define MD5_Update(ctx,data,len)	md5_update(ctx,len,data)
+#if NETTLE_VERSION_MAJOR >= 4
+#define MD5_Final(dig,ctx)	md5_digest(ctx,dig)
+#else
 #define MD5_Final(dig,ctx)	md5_digest(ctx,MD5_DIGEST_LENGTH,dig)
+#endif
 #else
 #endif
 
